### PR TITLE
🐛 ClusterClass: drop support for nullable

### LINKED ISSUE #5776: ClusterClass patches/variables: JSON `null` value does not work as expected with kubectl client-side apply
### Overview

Currently, we support nullable variables. This means that in the variable schema `nullable` can be set and then the corresponding variable on the Cluster can be also set to `null`.

Setting `null` as variable value works as expected. But setting `null` in `openAPIV3Schema.Default` and `jsonPatch.Value` doesn't work, i.e. setting them to `null` is treated the same way as not setting them at all.

### Details

We could modify our code to be able to **handle** `null` correctly by:
* adding `// +nullable` on the `Default` and `Value` fields so `null` is stored in etcd and sent to our webhook. 
* implementing custom `Unmarshal` functions for `JSONPatch` and `JSONSchemaProps` so our webhook can handle `null` values correctly.

These changes allow us to handle `null` correctly on the **server-side**, but there is also a severe issue with `kubectl apply`:

When using `kubectl apply` (with the default client-side apply), kubectl will try to unset the `null` value when the YAML surrounding `default` and `value` fields is changed. This happens because kubectl is using json merge patch. This is not an issue for the `value` in the Cluster resource but for `openAPIV3Schema.default` and `jsonPatch.value` on the ClusterClass. 

**NOTE**: server-side apply works as expected.

### Conclusion

We could improve the situation but we would still have a severe UX issue with kubectl apply when supporting `null`. We have the following options:

#### Option 1 (proposed): Drop support for nullable

Taking a few steps back, the question is if we have to support nullable:
* Optional variables are already possible by using a variable which is not required without a default value
    * Patches can then handle non-existing variables via `enabledIf` or a `valueFrom.template`.
* Fields on templates can also be removed with the `remove` JSON patch operation, we don't need the `null` value for that.

Based on that, it seems like the best option to drop support for `nullable` as there is no clear demand for it, and it's currently impossible to provide a good all around UX with `kubectl apply`. Also the changes we would have to make to support it fully (as outlined above)  are not breaking, so we could introduce support for `nullable` later on if we want to.

#### Option 2: Improve JSON `null` handling

In this case we would implement the changes as proposed above, but we have to live with the issue with `kubectl apply` and can only document the workaround of using server-side apply.

/kind bug
/area topology
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
More context on the issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5776

### CODE REVIEW

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 1]
```diff
@@ -83,9 +83,6 @@ func defaultClusterVariable(clusterVariable *clusterv1.ClusterVariable, clusterC
 
 	var value interface{}
 	// If the schema has a default value, default it.
-	// NOTE: This is essentially the same which the structural schema defaulting
-	// lib would do, except that it wouldn't default nullable values.
-	// As we want to also default nullable variables we have to do it ourselves.
 	if apiExtensionsSchema.Default != nil {
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:86]: Let's keep the simple defaulting for this PR. The new full implementation is part of: https://github.com/kubernetes-sigs/cluster-api/pull/5779 (including object/array support)

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 2]
```diff
@@ -132,8 +132,6 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, variableName string, fldP
 	switch {
 	case len(apiExtensionsSchema.Type) == 0:
 		return field.ErrorList{field.Required(fldPath.Child("type"), "openAPIV3Schema.type cannot be empty")}
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:134]: As we don't want to point to nullable anymore, it's good enough if `type == "null"` is handled by the next case.

**Comment by sbueringer** [comment]
Diff is fine, `Nullable` was never released.

**Comment by sbueringer** [comment]
/retest

flaky unrelated test

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**